subsys/net/lib/dns/resolve.c

```diff
@@ -1038,28 +1039,73 @@ int dns_validate_msg(struct dns_resolve_context *ctx,
 					ret = DNS_EAI_ADDRFAMILY;
 					goto quit;
 				}
+
+			} else if (ctx->queries[*query_idx].query_type ==
+							DNS_QUERY_TYPE_PTR) {
+				/* Synthesize a reply and place the returned info
+				 * in to the ai_canonname field. Use AF_LOCAL address
+				 * family as this is not a real address.
+				 */
+				struct net_buf *result;
+				uint8_t *pos;
+
+				address_size = MIN(dns_msg->response_length,
+						   DNS_MAX_NAME_SIZE);
+
+				/* Temporary buffer that is needed by dns_unpack_name()
+				 * to unpack the resolved name.
+				 */
+				result = net_buf_alloc(&dns_qname_pool, ctx->buf_timeout);
+				if (result == NULL) {
+					NET_DBG("Cannot allocate buffer for DNS query result");
+					ret = DNS_EAI_MEMORY;
+					goto quit;
+				}
+
+				pos = dns_msg->msg + dns_msg->response_position;
+				ret = dns_unpack_name(dns_msg->msg, dns_msg->msg_size, pos,
+						      result, NULL);
+				if (ret < 0) {
+					errno = -ret;
+					ret = DNS_EAI_SYSTEM;
+					net_buf_unref(result);
+					goto quit;
+				}
+
+				info.ai_family = AF_LOCAL;
+				info.ai_addrlen = MIN(result->len, DNS_MAX_NAME_SIZE);
+				memcpy(&info.ai_canonname, result->data, info.ai_addrlen);
+				info.ai_canonname[info.ai_addrlen] = '\0';
```

Comment:
I just wonder if we should document better (probably in `dns_resolve_service()` API description) how are results provided  to the callback in this case? I.e. that family will be set to AF_LOCAL, `ai_addrlen` would actually refer to length of data in `ai_canonname` and not `ai_addr` and that the actual result will be present in `ai_canonname` buffer. I think that could not be clear w/o peeking into the implementation.

Unless of course this is some standardized way I'm not familiar with.

Reply:
Very good point, I agree this needs to be documented. This is non standard and I tried to figure out a way to use the existing DNS API to get information about the services by using the `AF_LOCAL` address family. If we want to avoid using it like this, then we need to create a new API for this purpose.

Reply:
I think it's fine, just need some extra docs.